File: aqp/api/routes/service_manager.py
"""AQP service-manager API for local data-platform services."""
from __future__ import annotations

from typing import Any, Literal

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from aqp.services import service_manager

router = APIRouter(prefix="/service-manager", tags=["service-manager"])

ServiceName = Literal["trino", "polaris", "iceberg", "superset", "airbyte", "dagster", "neo4j"]
ActionName = Literal["start", "stop", "restart"]


class ServiceActionRequest(BaseModel):
    action: ActionName


class TrinoQueryRequest(BaseModel):
    statement: str
    catalog: str | None = None
    schema_: str | None = None


@router.get("/config")
def config() -> dict[str, Any]:
    return service_manager.config_snapshot()


@router.get("/health")
def health() -> dict[str, Any]:
    return service_manager.health()


@router.get("/{name}/health")
def single_health(name: ServiceName) -> dict[str, Any]:
    return service_manager.service_health(name)


@router.get("/{name}/logs")
def logs(
    name: ServiceName,
    lines: int = Query(default=200, ge=1, le=2000),
) -> dict[str, Any]:
    return service_manager.service_logs(name, lines=lines)


@router.post("/{name}/actions")
def action(name: ServiceName, payload: ServiceActionRequest) -> dict[str, Any]:
    result = service_manager.service_action(name, payload.action)
    if not result.get("ok") and result.get("enabled") is False:
        raise HTTPException(status_code=403, detail=result["error"])
    return result


# ---------------------------------------------------------------------------
# Iceberg / Polaris bootstrap
# ---------------------------------------------------------------------------


@router.get("/iceberg/status")
def iceberg_status() -> dict[str, Any]:
    return service_manager.iceberg_status()


@router.post("/iceberg/bootstrap")
def iceberg_bootstrap() -> dict[str, Any]:
    return service_manager.iceberg_bootstrap()


# ---------------------------------------------------------------------------
# Trino verification + recent queries
# ---------------------------------------------------------------------------


@router.post("/trino/verify")
def trino_verify() -> dict[str, Any]:
    return service_manager.trino_verify()


@router.get("/trino/queries")
def trino_queries(limit: int = Query(default=50, ge=1, le=200)) -> dict[str, Any]:
    return service_manager.trino_queries(limit=limit)


@router.post("/trino/query")
def trino_query(payload: TrinoQueryRequest) -> dict[str, Any]:
    return service_manager.trino_query(
        payload.statement, catalog=payload.catalog, schema=payload.schema_
    )

File: aqp_client/src/lib/api/serviceManager.ts
import { apiFetch } from "./client";

export type ServiceName =
  | "trino"
  | "polaris"
  | "iceberg"
  | "superset"
  | "airbyte"
  | "dagster"
  | "neo4j";

export interface ServiceHealth {
  ok: boolean;
  service?: string;
  error?: string | null;
  [key: string]: unknown;
}

export interface ServiceManagerHealth {
  ok: boolean;
  services: Record<ServiceName, ServiceHealth>;
  config: Record<string, unknown>;
}

export interface IcebergBootstrapStep {
  name: string;
  status: string;
  detail?: string;
  payload?: Record<string, unknown>;
}

export interface IcebergBootstrapReport {
  catalog: string;
  principal: string;
  principal_role: string;
  catalog_role: string;
  privilege: string;
  duration_seconds: number;
  success: boolean;
  bootstrap_required: boolean;
  steps: IcebergBootstrapStep[];
  credentials_file?: string | null;
  credentials_persisted: boolean;
  last_error?: string | null;
}

export interface IcebergStatus {
  catalog: string;
  principal: string;
  principal_role: string;
  catalog_role: string;
  privilege: string;
  polaris_reachable: boolean;
  catalog_present: boolean;
  principal_present: boolean;
  principal_role_present: boolean;
  catalog_role_present: boolean;
  credentials_persisted: boolean;
  credentials_file?: string;
  error?: string | null;
}

export interface TrinoVerification {
  coordinator_ok: boolean;
  coordinator_url: string;
  node_id?: string | null;
  node_version?: string | null;
  query_ok: boolean;
  iceberg_catalog_ok: boolean;
  catalogs?: string[];
  iceberg_schemas?: string[];
  error?: string | null;
}

export interface TrinoQueryRow {
  query_id: string;
  state: string;
  user?: string | null;
  source?: string | null;
  catalog?: string | null;
  schema?: string | null;
  elapsed_seconds?: number | null;
  queued_seconds?: number | null;
  error?: string | null;
  statement?: string;
  created?: string | null;
}

export const serviceManagerApi = {
  health: () => apiFetch<ServiceManagerHealth>("/service-manager/health"),
  logs: (name: ServiceName, lines = 200) =>
    apiFetch<{ ok: boolean; stdout?: string; stderr?: string; error?: string }>(
      `/service-manager/${name}/logs`,
      { query: { lines } },
    ),
  action: (name: ServiceName, action: "start" | "stop" | "restart") =>
    apiFetch<Record<string, unknown>>(`/service-manager/${name}/actions`, {
      method: "POST",
      body: JSON.stringify({ action }),
    }),
  icebergStatus: () => apiFetch<IcebergStatus>("/service-manager/iceberg/status"),
  icebergBootstrap: () =>
    apiFetch<IcebergBootstrapReport>("/service-manager/iceberg/bootstrap", { method: "POST" }),
  trinoVerify: () => apiFetch<TrinoVerification>("/service-manager/trino/verify", { method: "POST" }),
  trinoQueries: (limit = 50) =>
    apiFetch<{ queries: TrinoQueryRow[]; count: number }>("/service-manager/trino/queries", {
      query: { limit },
    }),
};
